unexpected stateClosing during HTTP request using wget

Using  `wget` to request a page from a version of your http-server example adapted for T1S,
I found that at the end of the wget communication a _Closing_ state is entered,
that will result in a panic within a `switch` statement in `ControlBlock.Recv`.

```
level=INFO msg=TCP:tx-statechange port=80 old=Established new=FinWait1 txflags=[FIN,ACK]
level=INFO msg=TCP:rx-statechange port=80 old=FinWait1 new=Closing rxflags=[FIN,ACK]
panic: unexpected stateClosing

goroutine 1 [running]:
github.com/soypat/seqs.(*ControlBlock).Recv(0x40000cc890, {0xf99d3916, 0x4b4, 0x0, 0xf8da, 0x10})
        github.com/soypat/seqs@v0.0.0-20240421220819-60c7db9451e0/control_user.go:187 +0x744
```

Apparently the condition `seg.ACK == tcb.snd.NXT` in `github.com/soypat/seqs@v0.0.0-20240421220819-60c7db9451e0/control.go:245` is not fulfilled in this case.

When using a browser instead of wget, I don't observe the Closing state: In this case, a direct
transition from FinWait1 to TimeWait can be seen:

```
level=INFO msg=TCP:rx-statechange port=80 old=FinWait1 new=TimeWait rxflags=[FIN,ACK]
```


I added the lines
```Go
	case StateClosing:
		if seg.Flags.HasAny(FlagACK) {
			tcb.state = StateTimeWait
			pending, err = FlagACK, nil
		}
```
-- based on a [TCP state diagram] -- to the switch statement in `ControlBlock.Recv`, which results in the following message flow:

[TCP state diagram]: https://upload.wikimedia.org/wikipedia/commons/f/f6/Tcp_state_diagram_fixed_new.svg

```
level=INFO msg=TCP:tx-statechange port=80 old=Established new=FinWait1 txflags=[FIN,ACK]
level=INFO msg=TCP:rx-statechange port=80 old=FinWait1 new=Closing rxflags=[FIN,ACK]
level=INFO msg=TCP:rx-statechange port=80 old=Closing new=TimeWait rxflags=[ACK]
level=INFO msg=lst:freeConnForReuse lport=80 rport=49158
```

Not sure if the added lines are sufficient.

Hi, please see the log -- obtained using my 10Base-T1S network -- below.

One difference I noticed when comparing wireshark capture logs is, that in the WLAN/cyw43439 setup (where the panic does not occur)  
-   first a [FIN,ACK] message appears to be sent from server to client (with seq. number 1191), setting the next sequence number to e.g. 1192,
-   then the client sends its [FIN,ACK] message, using an Ack value of 1192 that equals the next seq. number specified in the previous message.

On the T1S network, it is **the other way around**:
-   here the client sends a [FIN,ACK] msg with an Ack value of 1191,
-   then the server responds with a [FIN,ACK] msg

So when the server evaluates the if condition `seg.ACK == tcb.snd.NXT` in control.go:245, in case of the cyw43439 this will evaluate to `1192 == 1192`, while in case of my T1S-network it will evaluate to `1191 == 1192` which will make the condition fail.

Below the log with `slog.DEBUG-2`:
```
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=Stack.RecvEth:start plen=78
time=2024-05-05T14:46:07.646+02:00 level=DEBUG msg=TCP:recv opt=20 ipopt=0 payload=0
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=TCPConn.recv:start
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=tcb:rcv state=SynRcvd rcv.nxt=3774658309 rcv.wnd=1482 challenge=false
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=recv:seg seg.seq=3774658308 seg.ack=0 seg.wnd=64240 seg.flags=[SYN] seg.data=0
time=2024-05-05T14:46:07.646+02:00 level=INFO msg=TCP:rx-statechange port=80 old=Listen new=SynRcvd rxflags=[SYN]
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=TCPConn.stateCheck:hasPending
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=tcb:pending-out seg.seq=0 seg.ack=3774658309 seg.wnd=1482 seg.flags=[SYN,ACK] seg.data=0
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=tcb:snd state=SynRcvd pend=0 snd.nxt=1 snd.una=0 snd.wnd=64240
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=tcb:snd seg.seq=0 seg.ack=3774658309 seg.wnd=1482 seg.flags=[SYN,ACK] seg.data=0
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg=TCPConn.send:start
time=2024-05-05T14:46:07.646+02:00 level=DEBUG msg=TCP:send plen=54
time=2024-05-05T14:46:07.646+02:00 level=DEBUG-2 msg="Stack:\tHandleEth" plen=54
time=2024-05-05T14:46:07.657+02:00 level=DEBUG-2 msg=Stack.RecvEth:start plen=64
time=2024-05-05T14:46:07.657+02:00 level=DEBUG msg=TCP:recv opt=0 ipopt=0 payload=0
time=2024-05-05T14:46:07.657+02:00 level=DEBUG-2 msg=TCPConn.recv:start
time=2024-05-05T14:46:07.657+02:00 level=DEBUG-2 msg=tcb:rcv state=Established rcv.nxt=3774658309 rcv.wnd=1482 challenge=false
time=2024-05-05T14:46:07.657+02:00 level=DEBUG-2 msg=recv:seg seg.seq=3774658309 seg.ack=1 seg.wnd=64240 seg.flags=[ACK] seg.data=0
time=2024-05-05T14:46:07.658+02:00 level=INFO msg=TCP:rx-statechange port=80 old=SynRcvd new=Established rxflags=[ACK]
time=2024-05-05T14:46:07.658+02:00 level=DEBUG-2 msg=lst:noconn2snd
time=2024-05-05T14:46:07.658+02:00 level=INFO msg="new connection" remote=192.168.5.1:38102
time=2024-05-05T14:46:07.658+02:00 level=DEBUG-2 msg=TCPConn.SetReadDeadline:start
time=2024-05-05T14:46:07.658+02:00 level=DEBUG-2 msg=TCPConn.SetWriteDeadline:start
time=2024-05-05T14:46:07.658+02:00 level=DEBUG-2 msg=TCPConn.Read:start
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=Stack.RecvEth:start plen=184
time=2024-05-05T14:46:07.669+02:00 level=DEBUG msg=TCP:recv opt=0 ipopt=0 payload=126
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=TCPConn.recv:start
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=tcb:rcv state=Established rcv.nxt=3774658435 rcv.wnd=1482 challenge=false
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=recv:seg seg.seq=3774658309 seg.ack=1 seg.wnd=64240 seg.flags=[PSH,ACK] seg.data=126
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=TCPConn.stateCheck:hasPending
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=tcb:pending-out seg.seq=1 seg.ack=3774658435 seg.wnd=1356 seg.flags=[ACK] seg.data=0
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=tcb:snd state=Established pend=0 snd.nxt=1 snd.una=1 snd.wnd=64240
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=tcb:snd seg.seq=1 seg.ack=3774658435 seg.wnd=1356 seg.flags=[ACK] seg.data=0
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg=TCPConn.send:start
time=2024-05-05T14:46:07.669+02:00 level=DEBUG msg=TCP:send plen=54
time=2024-05-05T14:46:07.669+02:00 level=DEBUG-2 msg="Stack:\tHandleEth" plen=54
Got webpage request!
time=2024-05-05T14:46:07.674+02:00 level=DEBUG-2 msg=TCPConn.Write:start
time=2024-05-05T14:46:07.674+02:00 level=DEBUG-2 msg=TCPConn.Write:start
time=2024-05-05T14:46:07.680+02:00 level=DEBUG-2 msg=tcb:pending-out seg.seq=1 seg.ack=3774658435 seg.wnd=1482 seg.flags=[ACK] seg.data=1192
time=2024-05-05T14:46:07.680+02:00 level=DEBUG-2 msg=tcb:snd state=Established pend=0 snd.nxt=1193 snd.una=1 snd.wnd=64240
time=2024-05-05T14:46:07.681+02:00 level=DEBUG-2 msg=tcb:snd seg.seq=1 seg.ack=3774658435 seg.wnd=1482 seg.flags=[ACK] seg.data=1192
time=2024-05-05T14:46:07.681+02:00 level=DEBUG-2 msg=tcb:close state=Established
time=2024-05-05T14:46:07.681+02:00 level=DEBUG msg=TCP:delayed-close port=80
time=2024-05-05T14:46:07.681+02:00 level=DEBUG-2 msg=TCPConn.stateCheck:hasPending
time=2024-05-05T14:46:07.681+02:00 level=DEBUG-2 msg=TCPConn.send:start
time=2024-05-05T14:46:07.681+02:00 level=DEBUG msg=TCP:send plen=1246
time=2024-05-05T14:46:07.681+02:00 level=DEBUG-2 msg="Stack:\tHandleEth" plen=1246
time=2024-05-05T14:46:07.694+02:00 level=DEBUG-2 msg=Stack.RecvEth:start plen=64
time=2024-05-05T14:46:07.694+02:00 level=DEBUG msg=TCP:recv opt=0 ipopt=0 payload=0
time=2024-05-05T14:46:07.694+02:00 level=DEBUG-2 msg=TCPConn.recv:start
time=2024-05-05T14:46:07.694+02:00 level=DEBUG-2 msg=tcb:rcv state=Established rcv.nxt=3774658435 rcv.wnd=1482 challenge=false
time=2024-05-05T14:46:07.694+02:00 level=DEBUG-2 msg=recv:seg seg.seq=3774658435 seg.ack=1193 seg.wnd=63176 seg.flags=[ACK] seg.data=0
time=2024-05-05T14:46:07.695+02:00 level=DEBUG-2 msg=tcb:close state=Established
time=2024-05-05T14:46:07.695+02:00 level=DEBUG msg=TCP:delayed-close port=80
time=2024-05-05T14:46:07.695+02:00 level=DEBUG-2 msg=TCPConn.stateCheck:hasPending
time=2024-05-05T14:46:07.695+02:00 level=DEBUG-2 msg=tcb:pending-out seg.seq=1193 seg.ack=3774658435 seg.wnd=1482 seg.flags=[FIN,ACK] seg.data=0
time=2024-05-05T14:46:07.695+02:00 level=DEBUG-2 msg=tcb:snd state=FinWait1 pend=0 snd.nxt=1194 snd.una=1193 snd.wnd=63176
time=2024-05-05T14:46:07.695+02:00 level=DEBUG-2 msg=tcb:snd seg.seq=1193 seg.ack=3774658435 seg.wnd=1482 seg.flags=[FIN,ACK] seg.data=0
time=2024-05-05T14:46:07.695+02:00 level=INFO msg=TCP:tx-statechange port=80 old=Established new=FinWait1 txflags=[FIN,ACK]
time=2024-05-05T14:46:07.695+02:00 level=DEBUG-2 msg=TCPConn.send:start
time=2024-05-05T14:46:07.695+02:00 level=DEBUG msg=TCP:send plen=54
time=2024-05-05T14:46:07.695+02:00 level=DEBUG-2 msg="Stack:\tHandleEth" plen=54
time=2024-05-05T14:46:07.706+02:00 level=DEBUG-2 msg=Stack.RecvEth:start plen=64
time=2024-05-05T14:46:07.706+02:00 level=DEBUG msg=TCP:recv opt=0 ipopt=0 payload=0
time=2024-05-05T14:46:07.706+02:00 level=DEBUG-2 msg=TCPConn.recv:start
time=2024-05-05T14:46:07.706+02:00 level=DEBUG-2 msg=tcb:rcv state=Closing rcv.nxt=3774658436 rcv.wnd=1482 challenge=false
time=2024-05-05T14:46:07.706+02:00 level=DEBUG-2 msg=recv:seg seg.seq=3774658435 seg.ack=1193 seg.wnd=63176 seg.flags=[FIN,ACK] seg.data=0
time=2024-05-05T14:46:07.706+02:00 level=INFO msg=TCP:rx-statechange port=80 old=FinWait1 new=Closing rxflags=[FIN,ACK]
time=2024-05-05T14:46:07.706+02:00 level=DEBUG-2 msg=TCPConn.stateCheck:hasPending
time=2024-05-05T14:46:07.706+02:00 level=DEBUG-2 msg=Stack.RecvEth:start plen=64
time=2024-05-05T14:46:07.706+02:00 level=DEBUG msg=TCP:recv opt=0 ipopt=0 payload=0
time=2024-05-05T14:46:07.706+02:00 level=DEBUG-2 msg=TCPConn.recv:start
panic: unexpected stateClosing
```

--- a/control_user.go
+++ b/control_user.go
@@ -186,19 +186,19 @@ func (tcb *ControlBlock) Recv(seg Segment) (err error) {
 		if seg.Flags.HasAny(FlagACK) {
 			tcb.close()
 		}
-	// case StateClosing:
-	// 	if seg.Flags.HasAny(FlagACK) {
-	// 		tcb.state = StateTimeWait
-	// 		pending = FlagACK
-	// 	}
+	case StateClosing:
+		// Thanks to @knieriem for finding and reporting this bug.
+		if seg.Flags.HasAny(FlagACK) {
+			tcb.state = StateTimeWait
+		}
 	default:
-		panic("unexpected state" + tcb.state.String())
+		panic("unexpected recv state:" + tcb.state.String())
 	}
 	if err != nil {
 		return err
 	}
 
-	tcb.pending[0] = pending
+	tcb.pending[0] |= pending
 	if prevNxt != 0 && tcb.snd.NXT != prevNxt && tcb.logenabled(slog.LevelDebug) {
 		tcb.debug("tcb:snd.nxt-change", slog.String("state", tcb.state.String()),
 			slog.Uint64("seg.ack", uint64(seg.ACK)), slog.Uint64("snd.nxt", uint64(tcb.snd.NXT)),

--- a/seqs_test.go
+++ b/seqs_test.go
@@ -783,7 +783,15 @@ func TestIssue19(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
+	assertState(seqs.StateTimeWait)
 
+	// Check we still need to send an ACK.
+	pending, ok = tcb.PendingSegment(0)
+	if !ok {
+		t.Fatal("expected pending segment")
+	} else if pending.Flags != seqs.FlagACK {
+		t.Fatalf("expected ACK; got %s", pending.Flags.String())
+	}
 	// Prepare response to client.
 	err = tcb.Send(pending)
 	if err != nil {

--- a/seqs_test.go
+++ b/seqs_test.go
@@ -1,6 +1,8 @@
 package seqs_test
 
 import (
+	"log/slog"
+	"os"
 	"strconv"
 	"testing"
 
@@ -700,4 +702,91 @@ func TestUnexpectedStateClosing(t *testing.T) {
 		},
 	}
 	tcb.HelperExchange(t, ex[:])
+}
+
+// This corresponds to https://github.com/soypat/seqs/issues/19
+// The bug consisted of a panic condition encountered when using wget client with a seqs based server.
+// Thanks to @knieriem for finding this and the detailed report they submitted.
+func TestIssue19(t *testing.T) {
+	var tcb seqs.ControlBlock
+	tcb.SetLogger(slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
+		Level: slog.LevelDebug - 2,
+	})))
+	assertState := func(state seqs.State) {
+		t.Helper()
+		if tcb.State() != state {
+			t.Fatalf("want state %s; got %s", state.String(), tcb.State().String())
+		}
+	}
+	const httpLen = 1192
+	const issA, issB, windowA, windowB = 1, 0, 2000, 2000
+	tcb.HelperInitState(seqs.StateEstablished, issA, issA, windowA)
+	tcb.HelperInitRcv(issB, issB, windowB)
+
+	// Send out HTTP request and close connection.
+	err := tcb.Send(seqs.Segment{SEQ: issA, ACK: issB, Flags: PSHACK, WND: windowA, DATALEN: httpLen})
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = tcb.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertState(seqs.StateEstablished)
+
+	pending, ok := tcb.PendingSegment(0)
+	if !ok {
+		t.Fatal("expected pending segment")
+	} else if pending.Flags != FINACK {
+		t.Fatalf("expected FINACK; got %s", pending.Flags.String())
+	}
+
+	// Receive ACK of HTTP segment.
+	err = tcb.Recv(seqs.Segment{SEQ: issB, ACK: issA + httpLen, Flags: seqs.FlagACK, WND: windowB})
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertState(seqs.StateEstablished)
+	err = tcb.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	pending, ok = tcb.PendingSegment(0)
+	if !ok {
+		t.Fatal("expected pending segment")
+	} else if pending.Flags != FINACK {
+		t.Fatalf("expected FINACK; got %s", pending.Flags.String())
+	}
+
+	// Send out FINACK.
+	err = tcb.Send(pending)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertState(seqs.StateFinWait1)
+
+	// Receive FINACK response from client.
+	err = tcb.Recv(seqs.Segment{SEQ: issB, ACK: issA + httpLen, Flags: FINACK, WND: windowB})
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertState(seqs.StateClosing)
+	pending, ok = tcb.PendingSegment(0)
+	if !ok {
+		t.Fatal("expected pending segment")
+	} else if pending.Flags != seqs.FlagACK {
+		t.Fatalf("expected ACK; got %s", pending.Flags.String())
+	}
+
+	// Before responding we receive an ACK from client. This is where panic is triggered.
+	err = tcb.Recv(seqs.Segment{SEQ: issB + 1, ACK: issA + httpLen + 1, Flags: seqs.FlagACK, WND: windowB})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Prepare response to client.
+	err = tcb.Send(pending)
+	if err != nil {
+		t.Fatal(err)
+	}
 }

